From daf67e765bcfa6ff6affb1a916696271014303b2 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Mon, 21 Feb 2022 11:49:50 +0530
Subject: [PATCH] hostapd: add acs_exclude_6ghz_non_psc option for acs non
 offload driver

Hostapd configuration "acs_exclude_6ghz_non_psc" is supported for ACS
offloaded driver alone. Extend the support to ACS non offload driver
by adding 6G Preferred Scan Channel check and hostapd config flag
during scan and survey dump.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 src/ap/acs.c                 | 141 ++++++++++++++++++++++++++++++++++++++++---
 src/ap/ap_drv_ops.h          |   6 +-
 src/ap/bss_load.c            |   3 +-
 src/drivers/driver.h         |   4 +-
 src/drivers/driver_nl80211.c |  30 ++++++---
 5 files changed, 166 insertions(+), 18 deletions(-)

diff --git a/src/ap/acs.c b/src/ap/acs.c
index da13d48..61d0c4f 100644
--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -436,22 +436,6 @@ static bool acs_usable_bw_chan(const struct hostapd_channel_data *chan,
 	return false;
 }
 
-
-static int acs_get_bw_center_chan(int freq, enum bw_type bw)
-{
-	unsigned int i = 0;
-
-	while (bw_desc[bw][i].first != -1) {
-		if (freq >= bw_desc[bw][i].first &&
-		    freq <= bw_desc[bw][i].last)
-			return bw_desc[bw][i].center_chan;
-		i++;
-	}
-
-	return 0;
-}
-
-
 static int acs_survey_is_sufficient(struct freq_survey *survey)
 {
 	if (!(survey->filled & SURVEY_HAS_NF)) {
@@ -767,6 +751,112 @@ static void acs_update_puncturing_bitmap(struct hostapd_iface *iface,
 #endif /* CONFIG_IEEE80211BE */
 
 
+static void
+acs_find_6g_psc_chan_mode(struct hostapd_iface *iface,
+			  struct hostapd_hw_modes *mode,
+			  int n_chans, u32 bw,
+			  struct hostapd_channel_data **rand_chan,
+			  struct hostapd_channel_data **ideal_chan,
+			  long double *ideal_factor)
+{
+	struct hostapd_channel_data *chan, *adj_chan = NULL;
+	long double factor;
+	int i, j;
+	unsigned int k;
+
+	for (i = 0; i < mode->num_channels; i++) {
+		double total_weight;
+		struct acs_bias *bias;
+
+		chan = &mode->channels[i];
+		if (!chan_pri_allowed(chan))
+			continue;
+
+		if (!is_in_chanlist(iface, chan))
+			continue;
+
+		if (!is_in_freqlist(iface, chan))
+			continue;
+
+		if (!is_6ghz_psc_frequency(chan->freq))
+			continue;
+
+		if (!chan_bw_allowed(chan, bw, 1, 1)) {
+			wpa_printf(MSG_DEBUG,
+				   "ACS: Channel %d: BW %u is not supported",
+				   chan->chan, bw);
+			continue;
+		}
+
+		factor = 0;
+		if (acs_usable_chan(chan))
+			factor = chan->interference_factor;
+		total_weight = 1;
+
+		/* Start index from -1 because all the PSC channels are located in
+		 * secondary 20MHz of primary 40MHz. Ex: CH33 - CH61 has 160MHz BW, PSC
+		 * CH37 which is secondary 20MHz of primary 40MHz.
+		 */
+
+		for (j = -1; j < n_chans; j++) {
+			adj_chan = acs_find_chan(iface, chan->freq + (j * 20));
+			if (!adj_chan)
+				break;
+
+			if (!chan_bw_allowed(adj_chan, bw, 1, 0)) {
+				wpa_printf(MSG_DEBUG,
+					   "ACS: PRI Channel %d: secondary channel %d BW %u is not supported",
+					   chan->chan, adj_chan->chan, bw);
+				break;
+			}
+
+			if (acs_usable_chan(adj_chan)) {
+				factor += adj_chan->interference_factor;
+				total_weight += 1;
+			}
+		}
+
+		if (j != n_chans) {
+			wpa_printf(MSG_DEBUG, "ACS: Channel %d: not enough bandwidth",
+				   chan->chan);
+			continue;
+		}
+
+		factor /= total_weight;
+
+		bias = NULL;
+		if (iface->conf->acs_chan_bias) {
+			for (k = 0; k < iface->conf->num_acs_chan_bias; k++) {
+				bias = &iface->conf->acs_chan_bias[k];
+				if (bias->channel == chan->chan)
+					break;
+				bias = NULL;
+			}
+		}
+
+		if (bias) {
+			factor *= bias->bias;
+			wpa_printf(MSG_DEBUG,
+				   "ACS:  * channel %d: total interference = %Lg (%f bias)",
+				   chan->chan, factor, bias->bias);
+		} else {
+			wpa_printf(MSG_DEBUG,
+				   "ACS:  * channel %d: total interference = %Lg",
+				   chan->chan, factor);
+		}
+
+		if (acs_usable_chan(chan) &&
+		   (!*ideal_chan || factor < *ideal_factor)) {
+			*ideal_factor = factor;
+			*ideal_chan = chan;
+		}
+
+		/* This channel would at least be usable */
+		if (!(*rand_chan))
+			*rand_chan = chan;
+	}
+}
+
 static void
 acs_find_ideal_chan_mode(struct hostapd_iface *iface,
 			 struct hostapd_hw_modes *mode,
@@ -1056,10 +1146,18 @@ bw_selected:
 
 	for (i = 0; i < iface->num_hw_features; i++) {
 		mode = &iface->hw_features[i];
-		if (!hostapd_hw_skip_mode(iface, mode))
-			acs_find_ideal_chan_mode(iface, mode, n_chans, bw,
-						 &rand_chan, &ideal_chan,
-						 &ideal_factor);
+
+		if (!hostapd_hw_skip_mode(iface, mode)) {
+			if (iface->conf->acs_exclude_6ghz_non_psc) {
+				acs_find_6g_psc_chan_mode(iface, mode, n_chans, bw,
+							  &rand_chan, &ideal_chan,
+							  &ideal_factor);
+			} else {
+				acs_find_ideal_chan_mode(iface, mode, n_chans, bw,
+							 &rand_chan, &ideal_chan,
+							 &ideal_factor);
+			}
+		}
 	}
 
 	if (ideal_chan) {
@@ -1109,27 +1207,43 @@ static void acs_adjust_secondary(struct hostapd_iface *iface)
 
 static void acs_adjust_center_freq(struct hostapd_iface *iface)
 {
-	int center;
+	int psc_chan[] = {37, 53, 69, 85, 101, 117,
+			  133, 149, 165, 181, 197, 213} ;
+	int offset, i;
+	u8 bw = hostapd_get_oper_chwidth(iface->conf);
+	bool acs_exclude_6ghz_non_psc = iface->conf->acs_exclude_6ghz_non_psc;
+	bool is_sec_psc_chan = false;
 
 	wpa_printf(MSG_DEBUG, "ACS: Adjusting VHT center frequency");
 
-	switch (hostapd_get_oper_chwidth(iface->conf)) {
+	if (acs_exclude_6ghz_non_psc && (bw == CONF_OPER_CHWIDTH_160MHZ)) {
+		 for (i = 0; i < ARRAY_SIZE(psc_chan); i++) {
+			 if (psc_chan[i] == iface->conf->channel) {
+				 is_sec_psc_chan = (i%2) ? true : false;
+				 break;
+			 }
+		 }
+	 }
+
+	switch (bw) {
 	case CONF_OPER_CHWIDTH_USE_HT:
-		if (iface->conf->secondary_channel &&
-		    iface->freq >= 2400 && iface->freq < 2500)
-			center = iface->conf->channel +
-				2 * iface->conf->secondary_channel;
-		else if (iface->conf->secondary_channel)
-			center = acs_get_bw_center_chan(iface->freq, ACS_BW40);
-		else
-			center = iface->conf->channel;
+		offset = 2 * iface->conf->secondary_channel;
 		break;
 	case CONF_OPER_CHWIDTH_80MHZ:
-		center = acs_get_bw_center_chan(iface->freq, ACS_BW80);
-		break;
+		if (acs_exclude_6ghz_non_psc)
+			offset = 2;
+		else
+			offset = 6;
+		break;
 	case CONF_OPER_CHWIDTH_160MHZ:
-		center = acs_get_bw_center_chan(iface->freq, ACS_BW160);
-		break;
+		/* In 160MHz, if primary 20MHz present in secondary 80MHz, then
+		  * subtract with -6 to find the center frequency of the 160MHz
+		  */
+		if (acs_exclude_6ghz_non_psc)
+			offset = is_sec_psc_chan ? -6 : 10;
+		else
+			offset = 14;
+		break;
 	default:
 		/* TODO: How can this be calculated? Adjust
 		 * acs_find_ideal_chan() */
@@ -1138,7 +1252,8 @@ static void acs_adjust_center_freq(struct hostapd_iface *iface)
 		return;
 	}
 
-	hostapd_set_oper_centr_freq_seg0_idx(iface->conf, center);
+	hostapd_set_oper_centr_freq_seg0_idx(iface->conf,
+					     iface->conf->channel + offset);
 }
 
 
@@ -1238,7 +1353,8 @@ static void acs_scan_complete(struct hostapd_iface *iface)
 	wpa_printf(MSG_DEBUG, "ACS: Using survey based algorithm (acs_num_scans=%d)",
 		   iface->conf->acs_num_scans);
 
-	err = hostapd_drv_get_survey(iface->bss[0], 0);
+	err = hostapd_drv_get_survey(iface->bss[0], 0,
+				     iface->conf->acs_exclude_6ghz_non_psc);
 	if (err) {
 		wpa_printf(MSG_ERROR, "ACS: Failed to get survey data");
 		goto fail;
@@ -1288,6 +1404,11 @@ static int * acs_request_scan_add_freqs(struct hostapd_iface *iface,
 		if ((chan->flag & HOSTAPD_CHAN_INDOOR_ONLY) &&
 		    iface->conf->country[2] == 0x4f)
 			continue;
+  
+  		if (is_6ghz_freq(chan->freq) &&
+		     iface->conf->acs_exclude_6ghz_non_psc &&
+		     !is_6ghz_psc_frequency(chan->freq))
+			continue;
 
 		*freq++ = chan->freq;
 	}
diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index 331b0ea..19633f9 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -280,13 +280,15 @@ static inline void hostapd_drv_poll_client(struct hostapd_data *hapd,
 }
 
 static inline int hostapd_drv_get_survey(struct hostapd_data *hapd,
-					 unsigned int freq)
+					 unsigned int freq,
+					 bool acs_exclude_6ghz_non_psc)
 {
 	if (hapd->driver == NULL)
 		return -1;
 	if (!hapd->driver->get_survey)
 		return -1;
-	return hapd->driver->get_survey(hapd->drv_priv, freq);
+	return hapd->driver->get_survey(hapd->drv_priv, freq,
+					acs_exclude_6ghz_non_psc);
 }
 
 static inline int hostapd_get_country(struct hostapd_data *hapd, char *alpha2)
diff --git a/src/ap/bss_load.c b/src/ap/bss_load.c
index 725d3cd..f91493a 100644
--- a/src/ap/bss_load.c
+++ b/src/ap/bss_load.c
@@ -49,7 +49,8 @@ static void update_channel_utilization(void *eloop_data, void *user_data)
 	if (!(hapd->beacon_set_done && hapd->started))
 		return;
 
-	err = hostapd_drv_get_survey(hapd, hapd->iface->freq);
+	err = hostapd_drv_get_survey(hapd, hapd->iface->freq,
+				     hapd->iface->conf->acs_exclude_6ghz_non_psc);
 	if (err) {
 		wpa_printf(MSG_ERROR, "BSS Load: Failed to get survey data");
 		return;
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 70dab06..6e43add 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -4543,6 +4543,8 @@ struct wpa_driver_ops {
 	 * @priv: Private driver interface data
 	 * @freq: If set, survey data for the specified frequency is only
 	 *	being requested. If not set, all survey data is requested.
+	 * @acs_exclude_6ghz_non_psc: If set Include only preferred scan
+	 *	channels from 6 GHz band for ACS
 	 * Returns: 0 on success, -1 on failure
 	 *
 	 * Use this to retrieve:
@@ -4561,7 +4563,7 @@ struct wpa_driver_ops {
 	 * for each survey. The min_nf of the channel is updated for each
 	 * survey.
 	 */
-	int (*get_survey)(void *priv, unsigned int freq);
+	int (*get_survey)(void *priv, unsigned int freq, bool acs_exclude_6ghz_non_psc);
 
 	/**
 	 * status - Get driver interface status information
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 9dc0ed9..06b396d 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10029,6 +10029,10 @@ static int check_survey_ok(struct nlattr **sinfo, u32 surveyed_freq,
 	return freq_filter == surveyed_freq;
 }
 
+struct nl80211_get_survey_arg {
+	struct survey_results *survey_results;
+	bool acs_exclude_6ghz_non_psc;
+};
 
 static int survey_handler(struct nl_msg *msg, void *arg)
 {
@@ -10036,15 +10040,19 @@ static int survey_handler(struct nl_msg *msg, void *arg)
 	struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
 	struct nlattr *sinfo[NL80211_SURVEY_INFO_MAX + 1];
 	struct survey_results *survey_results;
+	struct nl80211_get_survey_arg *arg_survey =
+			(struct nl80211_get_survey_arg *)arg;
 	u32 surveyed_freq = 0;
 	u32 ifidx;
+	bool acs_exclude_6ghz_non_psc;
 
 	static struct nla_policy survey_policy[NL80211_SURVEY_INFO_MAX + 1] = {
 		[NL80211_SURVEY_INFO_FREQUENCY] = { .type = NLA_U32 },
 		[NL80211_SURVEY_INFO_NOISE] = { .type = NLA_U8 },
 	};
 
-	survey_results = (struct survey_results *) arg;
+	survey_results = arg_survey->survey_results;
+	acs_exclude_6ghz_non_psc = arg_survey->acs_exclude_6ghz_non_psc;
 
 	nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0),
 		  genlmsg_attrlen(gnlh, 0), NULL);
@@ -10073,6 +10081,11 @@ static int survey_handler(struct nl_msg *msg, void *arg)
 			     survey_results->freq_filter))
 		return NL_SKIP;
 
+	if (is_6ghz_freq(surveyed_freq) &&
+	    acs_exclude_6ghz_non_psc &&
+	    !is_6ghz_psc_frequency(surveyed_freq))
+		return NL_SKIP;
+
 	if (survey_results->freq_filter &&
 	    survey_results->freq_filter != surveyed_freq) {
 		wpa_printf(MSG_EXCESSIVE, "nl80211: Ignoring survey data for freq %d MHz",
@@ -10085,21 +10098,22 @@ static int survey_handler(struct nl_msg *msg, void *arg)
 	return NL_SKIP;
 }
 
-
-static int wpa_driver_nl80211_get_survey(void *priv, unsigned int freq)
+static int wpa_driver_nl80211_get_survey(void *priv, unsigned int freq,
+					 bool acs_exclude_6ghz_non_psc)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
 	struct nl_msg *msg;
 	int err;
 	union wpa_event_data data;
-	struct survey_results *survey_results;
+	struct nl80211_get_survey_arg arg;
 
 	os_memset(&data, 0, sizeof(data));
-	survey_results = &data.survey_results;
-
-	dl_list_init(&survey_results->survey_list);
+	os_memset(&arg, 0, sizeof(arg));
+	arg.survey_results = &data.survey_results;
+	arg.acs_exclude_6ghz_non_psc = acs_exclude_6ghz_non_psc;
 
+	dl_list_init(&arg.survey_results->survey_list);
 	msg = nl80211_drv_msg(drv, NLM_F_DUMP, NL80211_CMD_GET_SURVEY);
 	if (!msg)
 		return -ENOBUFS;
@@ -10110,7 +10124,7 @@ static int wpa_driver_nl80211_get_survey(void *priv, unsigned int freq)
 	do {
 		wpa_printf(MSG_DEBUG, "nl80211: Fetch survey data");
 		err = send_and_recv_msgs(drv, msg, survey_handler,
-					 survey_results, NULL, NULL);
+					 &arg, NULL, NULL);
 	} while (err > 0);
 
 	if (err)
@@ -10118,7 +10132,7 @@ static int wpa_driver_nl80211_get_survey(void *priv, unsigned int freq)
 	else
 		wpa_supplicant_event(drv->ctx, EVENT_SURVEY, &data);
 
-	clean_survey_results(survey_results);
+	clean_survey_results(arg.survey_results);
 	return err;
 }
 
